core/src/main/java/org/hibernate/ogm/jpa/impl/OgmJpaStoredProcedureQuery.java

```diff
@@ -0,0 +1,152 @@
+/*
+ * Hibernate OGM, Domain model persistence for NoSQL datastores
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.ogm.jpa.impl;
+
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+
+import javax.persistence.EntityManager;
+import javax.persistence.ParameterMode;
+import javax.persistence.StoredProcedureQuery;
+
+import org.hibernate.HibernateException;
+import org.hibernate.QueryParameterException;
+import org.hibernate.jpa.internal.StoredProcedureQueryImpl;
+import org.hibernate.jpa.spi.AbstractEntityManagerImpl;
+import org.hibernate.jpa.spi.HibernateEntityManagerImplementor;
+import org.hibernate.ogm.storedprocedure.impl.NoSQLProcedureCallImpl;
+import org.hibernate.ogm.storedprocedure.impl.NoSQLProcedureOutputImpl;
+import org.hibernate.ogm.util.impl.Log;
+import org.hibernate.ogm.util.impl.LoggerFactory;
+import org.hibernate.procedure.ParameterRegistration;
+import org.hibernate.procedure.ProcedureCallMemento;
+
+/**
+ * @author Sergey Chernolyas &amp;sergey_chernolyas@gmail.com&amp;
+ */
+public class OgmJpaStoredProcedureQuery extends StoredProcedureQueryImpl {
+
+	private static final Log log = LoggerFactory.make();
+	private Set<ParameterRegistration<?>> parameterRegistrations = new LinkedHashSet<>();
+	private NoSQLProcedureCallImpl procedureCall;
+	private HibernateEntityManagerImplementor entityManager;
+
+	public OgmJpaStoredProcedureQuery(NoSQLProcedureCallImpl procedureCall, EntityManager entityManager) {
+		super( procedureCall, convert( entityManager ) );
+		this.procedureCall = procedureCall;
+		this.entityManager = convert( entityManager );
+	}
+
+	public OgmJpaStoredProcedureQuery(ProcedureCallMemento memento, EntityManager entityManager) {
+		super( memento, convert( entityManager ) );
+	}
+
+	private static AbstractEntityManagerImpl convert(EntityManager em) {
+		if ( AbstractEntityManagerImpl.class.isInstance( em ) ) {
+			return (AbstractEntityManagerImpl) em;
+		}
+		throw new IllegalStateException( String.format( "Unknown entity manager type [%s]", em.getClass().getName() ) );
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public StoredProcedureQuery registerStoredProcedureParameter(int position, Class type, ParameterMode mode) {
+		entityManager().checkOpen( true );
+		parameterRegistrations.add( procedureCall.registerParameter( position, type, mode ) );
+		return this;
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	public StoredProcedureQuery registerStoredProcedureParameter(String parameterName, Class type, ParameterMode mode) {
+		entityManager().checkOpen( true );
+		parameterRegistrations.add( procedureCall.registerParameter( parameterName, type, mode ) );
```

Comment:
I'm surprised that you need to override these methods, in the end you are registering parameters couldn't you reuse the existing ones?

Reply:
I tried. Existing implementation coupled with JDBC. I need remove these links.

Reply:
Well see if you can update the other things, we can come back to this later